[ADD] barcode: operations and commands barcode printing (#13355)
odoo/documentation: content/applications/inventory_and_mrp/barcode/setup/operation_types.rst

============================
Operation types and commands
============================

Inventory operations, like :doc:`receipts and delivery orders
<../../inventory/shipping_receiving/daily_operations/receipts_delivery_one_step>`, as well as
commands, like validating a transfer or putting products in packages, can be performed by scanning
barcodes through the **Barcode** app. This allows employees to handle tasks and access menus without
the need to touch a screen or use the **Inventory** app on a computer, and allows workflow
streamlining by putting operation barcodes in the appropriate physical location.

.. image:: operation_types/barcode-default-operations.png
   :alt: Printable barcodes for Receipts, Delivery Orders, and Manufacturing.

.. _inventory/barcode/configure_operations:

Configuring operations
======================

*Operation types* in the **Inventory** and **Barcode** apps define actions that will change the

> state of products in inventory. Wich operation types are available to print will depend on the

Reviewer: Typo
```suggestion
state of products in inventory. Which operation types are available to print will depend on the
```

:menuselection:`Inventory app --> Configuration --> Operations Types` and tick the box by each
operation type required.

.. image:: operation_types/barcode-print-operations.png
   :alt: Available operations selected making a 'Print' button appear.

Next, click the :icon:`fa-print` :guilabel:`Print` button and select either :guilabel:`Operation
type (PDF)` to download a PDF with the operation barcodes, or :guilabel:`Operation type (ZPL)` to
send the print job to a :ref:`ZPL-enabled printer <iot/link-printer>`.

.. note::
   The :icon:`fa-print` :guilabel:`Print` button does not appear until at least one operation type
   is selected.

.. _inventory/barcode/print_order:

Print picking operations for specific orders
============================================

Receipts, manufacturing orders, transfers, and delivery orders can all be printed in advance and are
referred to in the print options menu as *picking operations* or *picking*.

In Barcode
----------

From the **Barcode** app, tap the :guilabel:`Operations` button, select the operation type, and then
the specific order to print. From there, either scan the `PRINT PICKING OPERATION` barcode if there

> is one available, or tap the :icon:`fa-cog` `(gear)` icon to open the **Barcode** actions menu and
> tap the :guilabel:`Print Picking Operation` button.

Reviewer: Please be mindful when [using find-and-replace regex](https://github.com/odoo/documentation/pull/13355#discussion_r2105199063), this still needs a guilabel
```suggestion
is one available, or tap the :icon:`fa-cog` :guilabel:`(gear)` icon to open the **Barcode** actions
menu and tap the :guilabel:`Print Picking Operation` button.
```

operation type required.

.. image:: operation_types/barcode-print-operations.png
   :alt: Available operations selected making a 'Print' button appear.

Next, click the :icon:`fa-print` :guilabel:`Print` button and select either :guilabel:`Operation
type (PDF)` to download a PDF with the operation barcodes, or :guilabel:`Operation type (ZPL)` to
send the print job to a :ref:`ZPL-enabled printer <iot/link-printer>`.

.. note::
   The :icon:`fa-print` :guilabel:`Print` button does not appear until at least one operation type
   is selected.

.. _inventory/barcode/print_order:

Print picking operations for specific orders
============================================

Receipts, manufacturing orders, transfers, and delivery orders can all be printed in advanced and
are referred to in the print options menu as "picking operations" or "picking".

In Barcode
----------

From the **Barcode** app, tap the :guilabel:`Operations` button, select the operation type, and
then the specific order to print. From there, either scan the :guilabel:`PRINT PICKING OPERATION`
barcode if there is one available, or tap the :icon:`fa-cog` :guilabel:`(cog)` icon to open the

> Barcode Actions menu and tap the :guilabel:`Print Picking Operation` button.

Author: But otherwise the rule of explicit label > hover label > fontAwesome name applies? This one's just an old convention?

.. image:: operation_types/print-picking-barcode-cog.png
   :alt: Menu from a receipt with the 'Print Picking Operation' button.

In Inventory
------------

To print a picking from the **Inventory** app, click the appropriate operation type
(*Receipt*, *Delivery Order*, etc.) and select the individual order or the
operation.

From the order, click the :icon:`fa-cog` :guilabel:`(Actions)` icon, hover over :icon:`fa-print`
:guilabel:`Print`, and click :guilabel:`Picking Operations`.

.. note::
   The type of operation can be inferred from the operation reference to help distinguish picking
   operations. For example, the "IN" in `WH/IN/00012` indicates that it is a receipt operation.
   These references are similar to, but do not exactly match the operation type barcodes.

Use barcodes for inventory operations and commands
==================================================

Operations
----------

> - :guilabel:`Receipts (WHIN)` creates a new order to receive products into inventory.

Author: Okay, I like this solution! I ran through italics/bold/literal for the whole phrase, but the parentheses are my addition as well. I'm going to convert it all to **Operation Name** `BARCODE`, because it looks the most like the actual printed barcodes, and the barcodes themselves seem to all stick to monospace fonts.

> - :guilabel:`Receipts (WHIN)` creates a new order to receive products into inventory.

Reviewer: Yes! I agree :)

\*\*Operation Name\*\* \`BARCODE\`

> - :guilabel:`Receipts (WHIN)` creates a new order to receive products into inventory.

Author: ![image](https://github.com/user-attachments/assets/492c8112-2ae2-4ff1-95e2-c91676a97ebd)

- :guilabel:`CANCEL`: Scan from an operation to prevent that operation from being validated and set
  its status to *Cancelled*.
- :guilabel:`PRINT PICKING OPERATION`: Scan from an existing delivery order, transfer, or receipt to
  generate a PDF with the name and barcode for that operation's reference number. This can be
  scanned later to go directly to the operation.
- :guilabel:`PRINT DELIVERY SLIP`: Scan from an existing receipt or delivery order to generate a PDF
  of the purchase order or delivery order. This will not include a barcode.
- :guilabel:`PUT IN PACK`: After scanning products, this command will designate them as all being in
  one :doc:`package <../../inventory/product_management/configure/package>`. Products scanned
  *after* this command are placed in a new package if the :guilabel:`PUT IN PACK` command is scanned
  again.
- :guilabel:`SCRAP`: Scan this to mark a scanned product as defective, and move it to a virtual
  location for :doc:`scrap inventory
  <../../inventory/warehouses_storage/inventory_management/scrap_inventory>`

Print barcodes for manufacturing commands
=========================================

> By default, the :guilabel:`Manufacturing` operation type can scan products and components and tap a
> :guilabel:`Produce` button to produce them. To print barcodes for basic manufacturing operations, go
> to :menuselection:`Manufacturing app --> Configuration --> Settings`, then, in the
> :guilabel:`Operations` section, make sure the :guilabel:`Barcode Scanner` box is ticked and click
> :icon:`fa-print` :guilabel:`Print barcode commands and operation types`.

Reviewer: No nav instructions yet, so please use italics here
```suggestion
By default, the *Manufacturing* operation type can scan products and components and tap a *Produce*
button to produce them. To print barcodes for basic manufacturing operations, go to
:menuselection:`Manufacturing app --> Configuration --> Settings`, then, in the
```

(*Receipt*, *Delivery Order*, etc.) and select the individual order or the
operation.

From the order, click the :icon:`fa-cog` :guilabel:`(Actions)` icon, hover over :icon:`fa-print`
:guilabel:`Print`, and click :guilabel:`Picking Operations`.

.. note::
   The type of operation can be inferred from the operation reference to help distinguish picking
   operations. For example, the "IN" in `WH/IN/00012` indicates that it is a receipt operation.
   These references are similar to, but do not exactly match the operation type barcodes.

Use barcodes for inventory operations and commands
==================================================

Operations
----------

- :guilabel:`Receipts (WHIN)` creates a new order to receive products into inventory.

   .. warning::
      Scanning a barcode for a receipt will always create a *new* instance of that operation, and
      will not match an already scheduled order. To scan or a specific scheduled receipt, first
      :ref:`print a picking for the individual receipt <inventory/barcode/print_order>`.

      .. example::
         Scanning :guilabel:`Receipts (WHIN)` will create a brand new receipt, even if the contents

>          match an existing receipt exactly.

Reviewer 2: very cool, and detailed! but I will once again complain about nested admonitions, but u do not have to change it :joy: 

![image](https://github.com/user-attachments/assets/6d5e3616-df77-4185-808f-1f72def7210e)

Reviewer 2: Can remove the example admonition and write it within the warning, like

(Also ignore the formatting since it's a mess)

" Scanning a barcode for a receipt will always create a *new* instance of that operation, and
      will not match an already scheduled order. To scan or a specific scheduled receipt, first
      :ref:`print a picking for the individual receipt <inventory/barcode/print_order>`.
      
      *Ex*: blabla in the example block

> - :guilabel:`Delivery Orders (WHOUT)` opens the currently scheduled delivery orders

Reviewer 2: ```suggestion
- :guilabel:`Delivery Orders (WHOUT)` opens the currently scheduled delivery orders.
```

  scanned later to go directly to the operation.
- :guilabel:`PRINT DELIVERY SLIP`: Scan from an existing receipt or delivery order to generate a PDF
  of the purchase order or delivery order. This will not include a barcode.
- :guilabel:`PUT IN PACK`: After scanning products, this command will designate them as all being in
  one :doc:`package <../../inventory/product_management/configure/package>`. Products scanned
  *after* this command are placed in a new package if the :guilabel:`PUT IN PACK` command is scanned
  again.
- :guilabel:`SCRAP`: Scan this to mark a scanned product as defective, and move it to a virtual
  location for :doc:`scrap inventory
  <../../inventory/warehouses_storage/inventory_management/scrap_inventory>`

Print barcodes for manufacturing commands
=========================================

By default, the :guilabel:`Manufacturing` operation type can scan products and components and tap a
:guilabel:`Produce` button to produce them. To print barcodes for basic manufacturing operations, go
to :menuselection:`Manufacturing app --> Configuration --> Settings`, then, in the

> :guilabel:`Operations` section, make sure the  :guilabel:`Barcode Scanner` box is ticked and click

Reviewer 2: ```suggestion
:guilabel:`Operations` section, make sure the :guilabel:`Barcode Scanner` box is ticked and click
```

> :icon:`fa-print` :guilabel:` Print barcode commands and operation types`.

Reviewer 2: ```suggestion
:icon:`fa-print` :guilabel:`Print barcode commands and operation types`.
```

> 

Reviewer 2: ```suggestion
```

Work order commands
-------------------

To enable more granular control of the manufacturing process through **Barcode**, enable the
:doc:`Work Orders <../../manufacturing/basic_setup/bill_configuration>` feature. To print barcodes
for work orders, go to :menuselection:`Manufacturing app --> Configuration --> Settings`, then, in
the :guilabel:`Operations` section, make sure the  :guilabel:`Work Orders` box is ticked and click
:icon:`fa-print` :guilabel:`Print barcode commands`.

.. image:: operation_types/print-work-order-commands.png

>    :alt: Add alt text.

Reviewer 2: remember to add alt text!

============================
Operation types and commands
============================

Inventory operations, like :doc:`receipts and delivery

> orders<../../inventory/shipping_receiving/daily_operations/receipts_delivery_one_step>`, as well

Reviewer 2: ```suggestion
orders <../../inventory/shipping_receiving/daily_operations/receipts_delivery_one_step>`, as well
```
we need one space between link text and link (:

commands, like validating a transfer or putting products in packages, can be performed by scanning
barcodes through the **Barcode** app. This allows employees to handle tasks and access menus without
the need to touch a screen or use the **Inventory** app on a computer, and allows workflow
streamlining by putting operation barcodes in the appropriate physical location.

.. image:: operation_types/barcode-default-operations.png
   :alt: Printable barcodes for Receipts, Delivery Orders, and Manufacturing.

.. _inventory/barcode/configure_operations:

Configuring operations
======================

*Operation types* in the **Inventory** and **Barcode** apps define actions that will change the
state of products in inventory. What operation types are available will depend on the settings which

> are enabled in the database. By default, barcodes can be printed for the :guilabel:`Receipts

Reviewer 2: ```suggestion
are enabled in the database. By default, barcodes can be printed for :guilabel:`Receipts
```
maybe an extra "the"? Also see my later comment about possibly changing these guilabels to italics if the operation types aren't visible on the screen (:

(WHIN)`, :guilabel:`Delivery Orders (WHOUT)`, and :guilabel:`Manufacturing (WHMANUF)`. The
:guilabel:`Internal Transfers (WHINT)` and :guilabel:`Internal Transfers (WHINT)` and

> :guilabel:`Pick (WHPICK)` operations can also be enabled.

Reviewer 2: all should change to italics or stay unformatted since we cant see them on the screen (:

.. note::
   Both :guilabel:`Internal Transfers (WHINT)` and :guilabel:`Pick (WHPICK)` operations require

>    :doc:`location barcodes <./software>` to scan.

Reviewer 2: ```suggestion
   :doc:`location barcodes <../software>` to scan.
```

Internal transfers
------------------

> The :guilabel:`Internal Transfers (WHINT)` operation barcode is available if the database is using
> the ::guilabel:`Storage Locations` to designate where products are located in the inventory. To

Reviewer 2: super nitpicky but we only use the guilabel tag when that field is currently visible on the UI. In this intro sentence, we haven't navigated to a specific page yet, so it's not yet on the reader's page. So, we use italics :blush: 

```suggestion
The *Internal Transfers (WHINT)* operation barcode is available if the database is using
the *Storage Locations* to designate where products are located in the warehouse. To
```

Note: extra `:` in front of storage locations, and maybe missing the word "feature"?

optional: I'm not really sure if "products are located in the inventory" is a common phrase? How about "in the warehouse", or "in inventory" (without "the")?

enable this setting, go to :menuselection:`Inventory app --> Configuration --> Settings`, then, in

> the :guilabel:`Warehouse` section, tick the :guilabel:`Storage Locations` option.

Reviewer 2: guilabels properly used here! :blush:

.. TODO: Seealso for transfers

Picking
-------

The :guilabel:`Pick (WHPICK)` operation barcode is available if the database is using the

> ::guilabel:`Multi-Step Routes` feature configured with a picking step. To enable this setting, go to

Reviewer 2: italics

:menuselection:`Inventory app --> Configuration --> Settings`, then, in the :guilabel:`Warehouse`
section, tick the :guilabel:`Multi-Step Routes` option and save. From there, click
:icon:`fa-arrow-right` :guilabel:`Set Warehouse Routes` to open warehouse settings and select the
warehouse. In the :guilabel:`Outgoing Shipments` section, select either :guilabel:`Pick then Deliver
(2 steps)` or :guilabel:`Pick, Pack, then Deliver (3 steps)`.

.. seealso::
   - :doc:`Two-step receipt and delivery
     <../../inventory/shipping_receiving/daily_operations/receipts_delivery_two_steps>`
   - :doc:`Three-step delivery
     <../../inventory/shipping_receiving/daily_operations/delivery_three_steps>`

> 

Reviewer 2: ```suggestion
```
extra line

> 

Reviewer 2: extra line
```suggestion
```

Print barcodes for inventory commands and operations
====================================================

Barcodes for operations and commands can be printed at any time. These documents will include all
the available inventory commands, regardless of whether they are enabled, but only the operations
that are currently enabled in the database. See :ref:`configuring operations
<inventory/barcode/configure_operations>` to enable any missing operations.

.. example::
   The `print_inventory_commands_and_operation_types` document will always show the :guilabel:`PUT
   IN PACK` command, regardless of whether the *Packages* feature is enabled, but the
   :guilabel:`Pick (WHPICK)` operation will only be added if *Multi-Step Routes* are configured to
   have a picking step.

.. warning::
   Printing barcodes, by default, will either directly download a PDF file of the selected barcodes
   or open a new tab with a downloadable or printable PDF. To turn this into a direct printing

>    operation, :doc:`connect a printer<../../../general/iot/devices/printer>`.

Reviewer 2: ```suggestion
   operation, :doc:`connect a printer <../../../general/iot/devices/printer>`.
```

   .. warning::
      Scanning a barcode for a receipt will always create a *new* instance of that operation, and
      will not match an already scheduled order. To scan or a specific scheduled receipt, first
      :ref:`print a picking for the individual receipt <inventory/barcode/print_order>`.

      .. example::
         Scanning :guilabel:`Receipts (WHIN)` will create a brand new receipt, even if the contents
         match an existing receipt exactly.

- :guilabel:`Delivery Orders (WHOUT)` opens the currently scheduled delivery orders
- :guilabel:`Manufacturing (WHMANUF)` creates a new manufacturing order. Scanning a product barcode
  from here will add it to the :guilabel:`Producing` list.
- :guilabel:`Internal Transfers (WHINT)` creates a new transfer order. :doc:`Location barcodes
  <software>` can be scanned to designate the source and destination, or scanned direction to
  initiate a transfer in **Barcode**.
- :guilabel:`Pick (WHPICK)` creates a new picking operation. This will *require* a :doc:`location
  barcode <software>` for the source location.

Commands
--------

- :guilabel:`MAIN MENU`: Scan to return to the main menu from an inventory adjustment.

> - :guilabel:`DISCARD`: TODO

Reviewer 2: let's delete this and all the other todos and revisit in follow-up PRs
```suggestion
```

- :guilabel:`VALIDATE`: Scan to confirm that an operation is correct and ready to be completed.
- :guilabel:`CANCEL`: Scan from an operation to prevent that operation from being validated and set

>   its status to "Canceled".

Reviewer 2: ```suggestion
  its status to *Canceled*.
```
